overlore/townhall/request.py

```diff
@@ -2,71 +2,120 @@
+    plotline: str = townhall_inputq if townhall_inputq != "" else townhall_db.fetch_plotline_by_realm(realm_id)
+    relevant_event = get_relevant_event(realm_id=realm_id, config=config)
 
-    # embed new discussion
-    embedding = llm.request_embedding(dialogue)
+    relevant_thoughts = get_npc_thoughts(realm_npcs, relevant_event, plotline, llm, townhall_db)
 
-    # insert our new discussion and its vector in our db
-    row_id = vector_db.insert_townhall_discussion(
-        discussion=dialogue,
-        summary=townhall["summary"],  # type: ignore[index]
-        realm_id=realm_id,
-        event_id=relevant_event_id,
-        embedding=embedding,
-    )
+    townhall = llm.generate_townhall_discussion(realm_name, realm_npcs, relevant_event, plotline, relevant_thoughts)
+    row_id = store_townhall_information(townhall, realm_id, realm_npcs, townhall_inputq, plotline, townhall_db, llm)
 
     return WsTownhallResponse(townhall_id=row_id, dialogue=townhall["dialogue"])  # type: ignore[index]
 
 
-def get_event_and_summary(
-    realm_id: int, vector_db: VectorDatabase, config: BootConfig
-) -> tuple[int, StoredEvent | None, str | None]:
+def get_relevant_event(realm_id: int, config: BootConfig) -> StoredEvent | None:
     events_db = EventsDatabase.instance()
     realms = Realms.instance()
 
     ts = get_katana_timestamp(config.env["KATANA_URL"])
 
-    relevant_event: StoredEvent | None = events_db.fetch_most_relevant_event(realms.position_by_id(realm_id), ts)
-
-    if relevant_event is None:
-        return (0, None, None)
-
-    relevant_event_id: int = cast(int, relevant_event[0])
-
-    summary = vector_db.get_townhall_from_event(relevant_event_id, realm_id=realm_id)
-
-    return (relevant_event_id, relevant_event, summary)
+    return events_db.fetch_most_relevant_event(realms.position_by_id(realm_id), ts)
 
 
 def convert_dialogue_to_str(dialogue: list[DialogueSegment]) -> str:
     return "\n".join(
         [f"{dialogue_segment['full_name']}: {dialogue_segment['dialogue_segment']}" for dialogue_segment in dialogue]  # type: ignore[index]
     )
+
+
+def get_npc_thoughts(
+    npc_list: list[NpcEntity],
+    relevant_event: StoredEvent | None,
+    plotline: str | None,
+    llm: Llm,
+    townhall_db: TownhallDatabase,
+) -> list[str]:
+    thoughts = []
+    query = ""
+
+    if relevant_event is not None:
+        query += llm.nl_formatter.event_to_nl(relevant_event)
+    if plotline is not None:
+        query += plotline
+    if query != "":
+        query_embedding = llm.request_embedding(query)
+        for npc in npc_list:
+            # townhall_db.query_nearest_neighbour
+            res = townhall_db.query_cosine_similarity(query_embedding, npc["entity_id"])
+            if res is not None:
+                row_id = res[0]
+                similarity = res[1]
+
+                # Arbitrary number, can be tuned accordingly
+                if similarity >= 0.3:
```

Comment:
So if no vector is found that is superior to 0.3 we don't include that thought, what does 0.3 represent ? Is there a scale somewhere that makes some sense of the number between 0 and 1 (0.1 is completely unrelated, 0.9 is very closely related or smthing)? 
Can we also put the 0.3 in a constant?